### PR TITLE
Enable continuous training for TextPredictor + Add paper references to the homepage

### PR DESCRIPTION
**Pull request by sxjscience**
*Description of changes:*

Enable to call `predictor.fit()` again after it's loaded or have been trained on a dataset.

Following are the examples of these two usages:

- Usage-1:

```python
from autogluon.text import TextPredictor
from autogluon.core.utils.loaders import load_pd

train_data = load_pd.load('https://autogluon-text.s3-accelerate.amazonaws.com/glue/sts/train.parquet')
dev_data = load_pd.load('https://autogluon-text.s3-accelerate.amazonaws.com/glue/sts/dev.parquet')

predictor = TextPredictor(label='score')
predictor.fit(train_data)
predictions1 = predictor.predict(dev_data)

# Call fit again to continue trainin
predictor.fit(train_data)
predictions2 = predictor.predict(dev_data)

```
- Usage-2:

```python
from autogluon.text import TextPredictor
from autogluon.core.utils.loaders import load_pd

train_data = load_pd.load('https://autogluon-text.s3-accelerate.amazonaws.com/glue/sts/train.parquet')
dev_data = load_pd.load('https://autogluon-text.s3-accelerate.amazonaws.com/glue/sts/dev.parquet')

predictor = TextPredictor(label='score', path='sts_model')
predictor.fit(train_data)
predictions1 = predictor.predict(dev_data)

# Load the model,  again to continue trainin
new_predictor = TextPredictor.load('sts_model')
new_predictor.fit(train_data, save_path='sts_model2')
predictions2 = new_predictor.predict(dev_data)
```

In addition, added the reference to the ICML workshop paper about multimodal text+tabular benchmark.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by szha** [comment]
Job PR-1299-5 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/5/index.html

**Comment by szha** [comment]
Job PR-1299-6 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/6/index.html

**Comment by szha** [comment]
Job PR-1299-9 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/9/index.html

**Comment by szha** [comment]
Job PR-1299-10 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/10/index.html

File: docs/tutorials/text_prediction/beginner.md [hunk 1]
```diff
@@ -140,6 +140,18 @@ for val, color in [(0, 'red'), (1, 'blue')]:
 plt.legend(loc='best')
 ```
 
+### Continuous Training
+:label:`sec_textprediction_continuous_training`
+
+You can also load a predictor and call `.fit()` again for continuous training.
```
Comment (jwmueller) [inline docs/tutorials/text_prediction/beginner.md:146]: user might not know what "for continuous training" means. May be clearer to say:
"to continue training the same predictor with additional data"

File: text/src/autogluon/text/text_prediction/predictor/predictor.py [hunk 2]
```diff
@@ -292,60 +318,87 @@ def fit(self,
...
+                raise AssertionError('No Text Column is found! This is currently not supported by '
+                                     'the TextPredictor. You may try to use '
+                                     'autogluon.tabular.TabularPredictor.\n'
+                                     'The inferred column properties of the training data is {}'
+                                     .format(column_types))
+            logger.info('Problem Type="{}"'.format(problem_type))
+            logger.info(printable_column_type_string(column_types))
+            self._problem_type = problem_type
+            if 'models' not in hyperparameters or 'MultimodalTextModel' not in hyperparameters['models']:
+                raise ValueError('The current TextPredictor only supports "MultimodalTextModel" '
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/predictor/predictor.py:368]: Do we need to clarify somewhere that MultimodalTextModel should be used even for data with only text? Otherwise users may find this confusing

File: text/src/autogluon/text/text_prediction/predictor/predictor.py [hunk 3]
```diff
@@ -292,60 +318,87 @@ def fit(self,
...
+            model_hparams = hyperparameters['models']['MultimodalTextModel']
+            self._backend = model_hparams['backend']
+            if plot_results is None:
+                plot_results = in_ipynb()
+
+            if self._backend == 'gluonnlp_v0':
+                import warnings
+                warnings.filterwarnings('ignore', module='mxnet')
+                from ..mx.models import MultiModalTextModel
+                self._model = MultiModalTextModel(column_types=column_types,
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/predictor/predictor.py:380]: The class is called `MultiModalTextModel` but the name in hyperparameters is `MultimodalTextModel`, I'd consider renaming the class to the latter as well to make them match.

File: text/src/autogluon/text/text_prediction/mx/models.py [hunk 4]
```diff
@@ -359,12 +364,20 @@ def train_function(args, reporter, train_df_path, tuning_df_path,
...
     log_metrics
         Metrics for logging
     eval_metric
         The stopping metric
     ngpus_per_trial
         The number of GPUs to use per each trial
+    params_path
+        The parameter path of the network
+    preprocessor_path
+        The path to store the preprocessed
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/mx/models.py:378]: preprocessor not preprocessed?

File: text/src/autogluon/text/text_prediction/predictor/predictor.py (hunk 3 above)
Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/predictor/predictor.py:380]: I'm not exactly sure if it will be backward compatible. Since we are also developing the PT version, we may later focus on the new version and keep the mxnet version unchanged.

File: text/src/autogluon/text/text_prediction/mx/models.py [hunk 5]
```diff
@@ -417,27 +429,37 @@ def train_function(args, reporter, train_df_path, tuning_df_path,
...
-        label_generator = None
-    preprocessor = MultiModalTextFeatureProcessor(column_types=column_types,
-                                                  label_column=label_column,
-                                                  tokenizer_name=cfg.model.backbone.name,
-                                                  label_generator=label_generator,
-                                                  cfg=cfg.preprocessing)
-    logger.info('Fitting and transforming the train data...')
-    train_dataset = preprocessor.fit_transform(train_data[feature_columns],
+    if continue_training:
+        with open(preprocessor_path, 'rb') as in_f:
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/mx/models.py:444]: maybe we want to print some statement here like:  "Continuing training of existing model"

File: text/src/autogluon/text/text_prediction/mx/models.py [hunk 6]
```diff
@@ -777,6 +802,34 @@ def get_recommended_resource(nthreads_per_trial=None,
...
+            if isinstance(cfg.optimization.optimizer_params[0], str):
+                fixed_optimizer_params = ast.literal_eval('[' + '. '.join(cfg.optimization.optimizer_params) + ']')
+                new_cfg.optimization.optimizer_params = fixed_optimizer_params
+        new_cfg.freeze()
+        return new_cfg
+
+
 @use_np
 class MultiModalTextModel:
     """Learner of the multimodal text data.
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/mx/models.py:835]: documentation for this model could be more specific, eg:  "This class implements a single neural network model that can operate on data with multiple text columns as well as tabular numeric/categorical columns."

File: text/src/autogluon/text/text_prediction/mx/models.py [hunk 7]
```diff
@@ -916,6 +974,8 @@ def train(self, train_data, tuning_data,
             algorithm, scheduling backend, HPO algorithm.
         search_space
             The search space options
+        continue_training
+            Whether we are loading a new model from scratch or we are continune model training
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/mx/models.py:978]: "continune model training" -> "continuing training of an existing model"

File: docs/tutorials/text_prediction/beginner.md (hunk 1 above)
Comment (sxjscience) [reply docs/tutorials/text_prediction/beginner.md]: Revised.

File: text/src/autogluon/text/text_prediction/mx/models.py (hunk 5 above)
Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/mx/models.py:444]: Added a logging message.

File: text/src/autogluon/text/text_prediction/mx/models.py [hunk 8]
```diff
@@ -982,15 +1043,27 @@ def train(self, train_data, tuning_data,
...
-                                      'cache_tuning_dataframe.pd.pkl')
+        # In fact, we may generalize this functionality to create the cache in S3/FSx.
+        cache_path = os.path.join(self._output_directory, 'cache')
+        os.makedirs(cache_path, exist_ok=True)
+        train_df_path = os.path.join(cache_path, 'cache_train_dataframe.pd.pkl')
+        tuning_df_path = os.path.join(cache_path, 'cache_tuning_dataframe.pd.pkl')
         train_data.to_pickle(train_df_path)
         tuning_data.to_pickle(tuning_df_path)
+        if continue_training:
+            # We need to store the current weights to the local disk as temporary cache.
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/mx/models.py:1057]: wouldn't the weights already be stored on disk in this case? I.e. wasn't the model loaded from disk before training is continued?

File: text/src/autogluon/text/text_prediction/mx/models.py (hunk 6 above)
Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/mx/models.py:835]: Added.

File: text/src/autogluon/text/text_prediction/mx/models.py (hunk 8 above)
Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/mx/models.py:1059]: It's possible that the predictor is called twice and the model hasn't been saved.

File: text/src/autogluon/text/text_prediction/predictor/predictor.py [hunk 9]
```diff
@@ -292,60 +318,87 @@ def fit(self,
...
-                                              feature_columns=feature_columns,
-                                              label_columns=label_columns,
-                                              problem_type=self._problem_type,
-                                              eval_metric=self._eval_metric,
-                                              log_metrics=log_metrics,
-                                              output_directory=self._path)
+        if is_continue_training:
+            # We have entered the continue training / transfer learning setting. The model is not None and should
+            # have been loaded by a previous `TextPredictor.load()` call.
+            logger.info('Calling .fit() for a model that has already been loaded. Start the transfer learning setting.')
```
Comment (jwmueller) [inline text/src/autogluon/text/text_prediction/predictor/predictor.py:325]: "Start the transfer learning setting." is not totally clear what it means.
Do you mean "Training will continue starting from the existing model rather than from scratch."

**Review by jwmueller — APPROVED** [review]
LGTM, just have minor comments and questions

**Comment by szha** [comment]
Job PR-1299-11 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/11/index.html

**Comment by szha** [comment]
Job PR-1299-13 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/13/index.html

**Comment by szha** [comment]
Job PR-1299-12 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/12/index.html

**Comment by szha** [comment]
Job PR-1299-14 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/14/index.html

**Comment by szha** [comment]
Job PR-1299-15 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/15/index.html

File: text/src/autogluon/text/text_prediction/predictor/predictor.py [hunk 10]
```diff
@@ -252,14 +256,38 @@ def fit(self,
...
         assert self._fit_called is False
+        is_continue_training = self._model is not None
         verbosity = self.verbosity
         if verbosity is None:
             verbosity = 3
-        if presets is not None:
-            preset_hparams = ag_text_presets.create(presets)
+        if save_path is not None:
+            self._path = setup_outputdir(save_path, warn_if_exist=True)
+        if is_continue_training:
```
Comment (Innixma) [inline text/src/autogluon/text/text_prediction/predictor/predictor.py:265]: Maybe add TODO: enforce version check. If predictor is from 0.3 and AG is installed from 0.4, this probably isn't a good idea -> may crash or have strange errors.

File: text/src/autogluon/text/text_prediction/mx/models.py [hunk 11]
```diff
@@ -1125,6 +1204,35 @@ def train(self, train_data, tuning_data,
...
+            clean_up_param_l.append(best_model_weight)
+        if cfg.model.use_avg_nbest:
+            clean_up_param_l.append(avg_nbest_path)
+
+        for ele in clean_up_param_l:
+            try:
+                os.remove(ele)
+            except OSError as e:
+                logger.info(f'Failed to remove the model weight from "{ele}".')
+
```
Comment (Innixma) [inline text/src/autogluon/text/text_prediction/mx/models.py:1235]: I'm afraid of bugs sneaking by in future changes because of the complexity of this cleaning. Is there any way for the model to keep track of all the cached files in an object so that instead of hard-coding all of these filenames it is instead fetched from an object? I'm ok with this being added as a TODO but it is something we should be mindful of, this logic makes it harder to work on this model code without adding bugs.

**Review by Innixma — APPROVED** [review]
Looks good! Just had a few comments.

File: text/src/autogluon/text/text_prediction/mx/models.py (hunk 11 above)
Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/mx/models.py:1235]: That's a valid concern. The appropriate approach might to create a temporary directory with unique id that stores everything inside and will be cleaned afterwards. I will address it in the later PR that adds pytorch.

Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/mx/models.py]: @Innixma I should have fixed the logic with uuid. In fact, we can later consider to create temporary space in S3. This will be useful for large datasets which cannot be stored in disk.

File: text/src/autogluon/text/text_prediction/predictor/predictor.py (hunk 10 above)
Comment (sxjscience) [reply text/src/autogluon/text/text_prediction/predictor/predictor.py:265]: It is currently backward compatible. Basically, we can load a predictor trained via AutoGluon 0.1--0.3 and continue train the model. The new model will be saved in the latest AutoGluon format.

**Comment by szha** [comment]
Job PR-1299-16 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/16/index.html

**Comment by szha** [comment]
Job PR-1299-17 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/17/index.html

**Comment by szha** [comment]
Job PR-1299-18 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-1299/18/index.html
